blog/2025-01-13-typescript-sdk-release.md

```diff
@@ -0,0 +1,375 @@
+const { embed } = post
+
+// We can now use the `$type` property to disambiguate
+if (embed?.$type === 'app.bsky.embed.images') {
+  // The `embed` variable is fully typed as `$Typed<AppBskyEmbedImages.Main>` here !
+}
+```
+
+### `is*` utility methods
+
+The example above shows how data can be discriminated based on the `$type` property. There are, however, several disadvantages to relying on string comparison for discriminating data types:
+
+- Having to use inline strings yields a lot of code, hurting readability and bundle size.
+- In particular instances, the `$type` property can actually have two values to describe the same lexicon. An "images" embed, for example, can use both `app.bsky.embed.images` and `app.bsky.embed.images#main` as `$type`. This makes the previous point even worse.
+
+In order to alleviate these issues, the SDK provides type checking predicate functions. In their previous implementation, the `is*` utilities were defined as follows:
+
+```typescript
+export interface Main {
+  images: Image[]
+  [x: string]: unknown
+}
+
+export function isMain(value: unknown): values is Main {
+  return (
+    value != null &&
+    typeof value === 'object' &&
+    '$type' in value &&
+    (value.$type === 'app.bsky.embed.images' ||
+      value.$type === 'app.bsky.embed.images#main')
+  )
+}
+```
+
+As can be seen from the example implementation above, the predicate functions would cast any object containing the expected `$type` property into the corresponding type, without checking for the actual validity of other properties. This could yield runtime errors that could have been avoided during development:
+
+```typescript
+import { AppBskyEmbedImages } from '@atproto/api'
+
+// Alias, for clarity
+const isImages = AppBskyEmbedImages.isMain
+
+// Get an invalid embed somehow
+const invalidEmbed = {
+  $type: 'app.bsky.embed.images',
+  // notice how the `images` property is missing here
+}
+
+// This predicate function only checks the value of the `$type` property, making the condition "true" here
+if (isImages(invalidEmbed)) {
+  // No TypeScript error, BUT causes a runtime error because there is no "images" property !
+  console.log('First image:', invalidEmbed.images[0])
+}
+```
+
+The root of the issue here is that the `is*` utility methods perform type casting of objects solely based on the value of their `$type` property. There were basically two ways of fixing this issue:
+
+1. Alter the implementation to actually validate the object's structure. This is a non-breaking change that has a negative impact on performance.
+2. Alter the function signature to describe what the function actually does. This is a breaking change because TypeScript would start (rightfully) returning lots of errors in places where these functions are used.
+
+Because this release introduces other breaking changes, and because adapting our own code base to this change showed it made more sense, we decided to adopt the latter option.
```

Comment:
```suggestion
Because this release introduces other breaking changes, and because adapting our own codebase to this change showed it made more sense, we decided to adopt the latter option.
```